### ISSUE TITLE
Why using replay_buffer.gather_all in PPO?

### ISSUE THREAD

**Issue by ChengshuLi** [tensorflow/agents#79]
Hi,

The PPO code seems to be using replay_buffer.gather_all()
https://github.com/tensorflow/agents/blob/748bede752a95aee4d73c2fd208202bcdf164e0e/tf_agents/agents/ppo/examples/v1/train_eval.py#L213

whereas DDPG and TD3 seem to be using replay_buffer.as_dataset
https://github.com/tensorflow/agents/blob/442e05f1da8a591306ede197f402295c4bdee003/tf_agents/agents/td3/examples/v1/train_eval.py#L194-L197

My feeling is that we should almost always use replay_buffer.as_dataset since if our replay_buffer is large, then using replay_buffer.gather_all (i.e. using all the experiences in one batch) won't fit into the gpu memory. Am I missing something? Thanks a lot in advance.

**Comment by samfishman** [comment]
gather_all is necessary because the PPO algorithm is designed to train on entire batch of on-policy data at once. Since it only gathers a small number of episodes at a time (think 30-120 episodes), trains on them, then clears the buffer and repeats, there is never too much data to train on at once.

**Comment by ChengshuLi** [comment]
@samfishman 
I see. What if I am using images as my observation? Using a batch size of 1000 or 10000 for images overflows my gpu memory.

**Comment by samfishman** [comment]
What do you mean by batch size in this context? I would suggest starting by just lowering the number of steps you collect per iteration to the largest size that fits in memory.

**Comment by ChengshuLi** [comment]
@samfishman 

Your suggestion makes sense. 

Just to clarify:
Say my replay buffer size is 10000.
I collect 20 episodes per iteration, each episode has 500 steps.
This will exactly fill up the replay buffer because 20 * 500 = 10000

Say my observation space is an image of size (128, 128, 3)
Calling `replay_buffer.gather_all` and `agent.train(trajectories)` will feed a tensor of size (10000, 128, 128, 3) through my network. I think this is what I mean by "batch size" of 10000. In this case, it will overflow my gpu memory. 

Your suggestion is to reduce `collect_episodes_per_iteration`, the number 20 in my example above, correct? 

Thank you very much.

**Comment by samfishman** [comment]
Yep, your understanding of my suggestion is correct. Thanks for clarifying what you mean by "batch size" here. Unfortunately there's not really a great alternative here aside from buying a GPU with more memory or using a cloud GPU service (such as GCP or AWS) to gain access to a large GPU or pool of GPUs you can distribute over.

**Comment by ChengshuLi** [comment]
Awesome. Thanks a lot!

**Comment by dimitry12** [comment]
> gather_all is necessary because the PPO algorithm is designed to train on entire batch of on-policy data at once

@samfishman I am still confused. I can see how advantage-estimation step benefits from looking at the whole training episode, but my understand of PPO is that backprop on value/actor-network can be done using mini-batches.

See https://github.com/openai/baselines/blob/c02b575f017a4dacceb8d98d1d5a16a21480faca/baselines/ppo2/ppo2.py#L143 for example.

**Comment by bruno-dasilva** [comment]
Following up, hoping to hear some input as well - from my understanding, PPO paper design and implementations (see baselines link above) typically gather a large set of experiences/steps/trajectories and then apply multiple rounds of SGD in smaller batches sampled from that set.

So, for example, you could obtain 10,000 steps of trajectories (buffer_size) for a single training iteration. Then at training time, you partition your trajectories into random groups of 64 (batch_size) samples and run minibatch SGD over each batch. Repeat this 30 (num_epochs) times. Note that gradients are updated within each minibatch run (ie not saved and run all at once later). Then the trajectories are cleared and it starts again.

Outside of the training and convergence benefits of mini batching, this conveniently lets you reduce how much needs to be on the GPU at any one time. When you're only passing in a tensor of size (64, 128, 128, 3) for a batch size of 64, it greatly reduces the in-memory requirement. This is particularly true with larger networks or larger inputs. In my experience, it's not uncommon to see gpu memory usage nearing 10GB for a minibatch size of 128 in larger networks.
(quick math: if you have 10000 trajectories your input tensor is float32 of shape (10000,128,128,3), the size of **that tensor alone** is `10000*128*128*3*4` = 1966080000 bytes, or **1.8 GB**.)

Now, I can't say I fully and intuitively comprehend the PPO algorithm at a low enough level to really grasp the importance minibatch here for training. But it seems like an important performance consideration -- from my understanding, mini batching actually -improves- convergence. Given we don't have any official benchmarks, it's hard to say what degree of improvement that would be. So, then, I'm curious - @samfishman is there some reason (performance or otherwise?) why we're choosing to not minibatch here?

**Comment by mattinjersey** [comment]
could you explain replay_buffer_size. If there are 1000 observations per step, in 50 steps should the replay buffer size be 50 or should it be 50,000.

**Comment by sguada** [comment]
So once you have computed the returns then you can train in batches and don't need the full buffer.
Currently we don't a way to extract the reward from the buffer, compute the returns and then sample mini-batches. But seems a good thing to add.

**Comment by ebrevdo** [comment]
We're no longer using gather_all; we instead either do minibatch PPO or reverb as_dataset.
